subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/swift/SwiftCompiler.java

```diff
@@ -71,7 +72,7 @@ protected void addOptionsFileArgs(List<String> args, File tempDir) {
 
     @Override
     public WorkResult execute(SwiftCompileSpec spec) {
-        if (swiftCompilerVersion.getMajor() < spec.getSourceCompatibility().getVersion()) {
+        if (swiftCompilerVersion.getMajor() < spec.getSourceCompatibility().getVersion() || (swiftCompilerVersion.getMajor() >= 5 && spec.getSourceCompatibility().equals(SwiftVersion.SWIFT3))) {
```

Comment:
WDYT about having a `SwiftVersion.isSupported(SwiftVersion)` where Swift 3 can support Swift 3, Swift 4 can support Swift 3&4 and Swift 5 can support Swift 4&5.

Reply:
It's a great idea, but I'm unsure we want the method to appear in the public API, anyway not yet. Alternatively, we could convert the `SwiftVersion` into an interface so we can have an internal API.